cumulusci/utils/yaml/cumulusci_yml.py

```diff
@@ -1,4 +1,276 @@
-cci_safe_load = load_yaml_data
+from typing_extensions import Literal, TypedDict
+from pydantic import Field
+
+from cumulusci.utils.fileutils import DataInput, load_from_source
+from cumulusci.utils.yaml.model_parser import CCIDictModel
+from cumulusci.utils.yaml.safer_loader import load_yaml_data
+
+default_logger = getLogger(__name__)
+
+
+NBSP = "\u00A0"
+
+pattern = re.compile(r"^\s*[\u00A0]+\s*", re.MULTILINE)
+
+
+#  type aliases
+PythonClassPath = str
+URL = str
+
+
+class FlowReference(CCIDictModel):
+    flow: str = None
+    options: Dict[str, Any] = {}
+    ignore_failure: bool = None  # is this allowed?
+    when: str = None  # is this allowed?
+    ui_options: Dict[str, Any] = {}
+
+
+class TaskReference(CCIDictModel):
+    task: str
+    options: Dict[str, Any] = {}
+    ui_options: Dict[str, Any] = {}
+    ignore_failure: bool = None
+    when: str = None  # is this a documented feature?
+    # classpath, description allowed?
+
+
+class Task(CCIDictModel):
+    name: str = None  # is this real or a typo
+    class_path: str = None
+    description: str = None
+    options: Dict[str, Any] = None
+    group: str = None
+    ui_options: Dict[str, Any] = None
+
+
+class Flow(CCIDictModel):
+    description: str = None
+    steps: Dict[str, Union[FlowReference, TaskReference]]
+    group: str = None
+
+
+class Package(CCIDictModel):
+    name: Optional[str] = None
+    name_managed: Optional[str] = None
+    namespace: Optional[str] = None
+    install_class: str = None
+    uninstall_class: str = None
+    api_version: str = None
+
+
+class Test(CCIDictModel):
+    name_match: str
+
+
+class Parser(CCIDictModel):
+    class_path: PythonClassPath
+    title: str
+
+
+class ReleaseNotes(CCIDictModel):
+    parsers: Dict[int, Parser]  # should probably be a list
+
+
+class Git(CCIDictModel):
+    repo_url: str = None
+    default_branch: str = None
+    prefix_feature: str = None
+    prefix_beta: str = None
+    prefix_release: str = None
+    push_prefix_sandbox: str = None
+    push_prefix_production: str = None
+    release_notes: ReleaseNotes = None
+    two_gp_context: str = Field(None, alias="2gp_context")
+
+
+class ApexDoc(CCIDictModel):
+    homepage: str = None  # can't find these in the CCI Python code
+    banner: str = None
+    branch: str = None
+    repo_dir: Path = None
+
+
+class Check(CCIDictModel):
+    when: str = None
+    action: str = None
+    message: str = None
+
+
+class Plan(CCIDictModel):  # MetaDeploy plans
+    title: str
+    description: str = None
+    tier: str = None
+    slug: str = None
+    is_listed: bool = None
+    steps: Dict[str, Union[FlowReference, TaskReference]]
+    checks: List[Check] = None
+    group: str = None
+    error_message: str = None
+    error_message: str = None
+    post_install_message: str = None
+    preflight_message: str = None
+
+
+class Project(CCIDictModel):
+    name: Optional[str]
+    package: Package = None
+    test: Test = None
+    git: Git = None
```

Comment:
We don't need to make these optional if they have values in the universal config, right?

Reply:
I think it is more helpful to the user if we validate each file by itself. How can we tell them where their error is if the "missing value" could be anywhere in four files.

Reply:
I think we'll need to add additional methods to each model to do an extra "semantic" validation after merger. That would be a separate WI.

Reply:
> How can we tell them where their error is if the "missing value" could be anywhere in four files.

We fundamentally can't; we can just tell them what is missing and let them decide what scope is most appropriate to add it (likely the project config).

But I suppose you're thinking more of errors that are wrong types rather than missing values, and that it's better if we can tell the user which file and line the problem is at. That's true but I'm not sure it's worth creating models that incorrectly declare fields as optional. I wonder if we need the dictmerge to construct a sort of sourcemap so that we can identify where each value came from (does the yaml parser even give us a way to preserve info about lines?)

Reply:
It isn't incorrect in saying that those values are optional. They are optional in the "CumulusCI.yml file format" because the can be specified elsewhere.

The Python "file format parser" allows you to call len() with zero arguments, but the runtime will complain that you've left something out. The parser isn't wrong: it's just responsible for a particular set of concerns.

Sourcemap seems very complicated and providing very little benefit over doing it the way I suggest. You can't catch every error at parse time and I don't see that as a problem even a little bit. You can do a semantic review later with dramatically less effort than trying to build a sourcemap for Python primitive objects.

"Task foo does not have a class_path" doesn't tie back to a particular source file anyhow.

I did the sourcemap thing in Snowfakery because the benefit there was large but as you probably remember and have noticed, line number handling had to suffuse the whole code base. It was a lot of work and I don't see the cost/benefit the same way in CumulusCI. We'd need to identify WHICH errors can only be found after merge AND validate that they actually even relate to a single file as opposed to the project as a whole.

As an aside: there's a part of me that thinks that dictmerge may be a clever hack that has outlived its usefulness. As you know I have another way I prefer for projects to inject steps and if we enumerate every dict-merge use-case maybe we'd find that dedicated features work better for all of them. Just a guess and an aside.

Reply:
In that sense *everything* is optional in cumulusci.yml. But I don't want to have to maintain two different models of our configuration, one which pretends everything is optional and one which actually indicates what is required at runtime. I think we want one model that indicates what's required at runtime, and then figure out how to ignore missing required items while parsing a specific file if we need to do it that way.

I agree a sourcemap is complicated and don't want to go that way unless we have to. Just throwing ideas out.

I also don't like dictmerge much, but I care a lot about not forcing users to rewrite their existing configuration files, so changing it feels daunting.

Reply:
"But I don't want to have to maintain two different models of our configuration, one which pretends everything is optional and one which actually indicates what is required at runtime."

I don't think we need two models. We need two algorithms. We can have one algorithm driven by Pydantic that looks primarily for types and extra keys and another driven by Python code that validates various semantic aspects, including the presence of required-at-runtime keys.

"I think we want one model that indicates what's required at runtime, and then figure out how to ignore missing required items while parsing a specific file if we need to do it that way."

It seems far easier to me to layer on an additional check than to "turn off" a required item
I know how to write that code straightforwardly.

In fact, I wrote that kind of code at your request:

https://github.com/SFDO-Tooling/CumulusCI/pull/1624/files#diff-66f8c248cf465233668398ce87f73e4cf911807618fa87b39f5224a39d67fe8fR29

In that case, I had accurately expressed true "requirements" at the syntactic Iayer and I switched it to be a post-validation at your request because I don't see why it is important that the schema be "accurate" as long as it generates the right error messages for the user.

I literally don't have a good idea about how to tell Pydantic: "this string is required except when it isn't." I can't imagine that they would be open to such a feature request.

I am sure that with a day or two of programming there is SOME way, but it seems like a lot of complication for not much benefit to me.

If I understand you, you want something like this:

```
class Task(CCIDictModel):
    class_path: OptionalAtParseTimeRequiredAtRuntime[str] = None
```

And I want something like this:

```
class Task(CCIDictModel):
    class_path: str = None
    
    def runtime)check(self):
        if not self.class_path:
            raise ValidationError("No class_path")
```

Reply:
No, I want this:

```
class Task(CCIDictModel):
    class_path: str
```

because that's the type that accurately describes the fact that we want to write code which can assume the presence of a class_path once it has a validated task configuration. I'm anticipating that we'll eventually want to use these models in type annotations in the codebase, not just for validation, and I don't want my type checker to tell me I need to handle None everywhere.

And then I want 2 different kinds of validation while loading the configuration:
1. A _partial_ validation of each file being loaded (i.e. validate the keys that are present to make sure they match the types from the model, but don't yet check that all required fields are there) so we can indicate what file has the wrong type.
2. A full validation of the merged configuration, so that we can catch missing required fields -- without needing to write custom validators.

https://github.com/samuelcolvin/pydantic/issues/1673 has some discussion of doing partial validation, with a suggestion from Samuel to use `validate_model` directly.